File: cpp/tempconv.cpp
#include <iostream>
#include <iomanip>

class Celsius
{
public:
  double value;
  Celsius(double v) : value(v) {}
  void print() const
  {
    std::cout << std::fixed << std::setprecision(1) << value << "°C\n";
  }
};

class Fahrenheit
{
public:
  double value;
  Fahrenheit(double v) : value(v) {}
  void print() const
  {
    std::cout << std::fixed << std::setprecision(1) << value << "°F\n";
  }
};

const Celsius AbsoluteZeroC(-273.15);
const Celsius FreezingC(0);
const Celsius BoilingC(100);

Fahrenheit CToF(const Celsius &c)
{
  return Fahrenheit(c.value * 9 / 5 + 32);
}

Celsius FToC(const Fahrenheit &f)
{
  return Celsius((f.value - 32) * 5 / 9);
}

int main()
{
  Celsius c(25);
  Fahrenheit f = CToF(c);
  f.print();

  f = Fahrenheit(77);
  c = FToC(f);
  c.print();

  return 0;
}

File: typescript/tempconv.ts
export class Celsius {
  constructor(public value: number) { }
  toString(): string {
    return `${this.value.toFixed(1)}°C`;
  }
}

export class Fahrenheit {
  constructor(public value: number) { }
  toString(): string {
    return `${this.value.toFixed(1)}°F`;
  }
}

export const AbsoluteZeroC = new Celsius(-273.15);
export const FreezingC = new Celsius(0);
export const BoilingC = new Celsius(100);

export function CToF(c: Celsius): Fahrenheit {
  return new Fahrenheit(c.value * 9 / 5 + 32);
}

export function FToC(f: Fahrenheit): Celsius {
  return new Celsius((f.value - 32) * 5 / 9);
}
